pkg/util/io/io.go

```diff
@@ -20,12 +20,114 @@ import (
 	"fmt"
 	"io/ioutil"
 	"os"
+	"path/filepath"
+	"time"
 
 	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/apimachinery/registered"
 	"k8s.io/kubernetes/pkg/runtime"
 )
 
+// ioTimeout is the timeout to use for io operations
+const ioTimeout time.Duration = 5
+
+// Lstat wraps os.Lstat in a timeout.
+func Lstat(path string) (os.FileInfo, error) {
+	return internalLstatTimeout(path, ioTimeout)
+}
+
+func internalLstatTimeout(path string, timeout time.Duration) (os.FileInfo, error) {
+	var (
+		timer      = time.NewTimer(ioTimeout)
+		resChannel = make(chan os.FileInfo)
+		errChannel = make(chan error)
+	)
+
+	defer timer.Stop()
+
+	go func() {
+		fileInfo, err := os.Lstat(path)
+		if err != nil {
+			errChannel <- err
+		} else {
+			resChannel <- fileInfo
+		}
+	}()
+
+	select {
+	case <-timer.C:
+		return nil, fmt.Errorf("Lstat on %v timed out after %v", path, timeout)
+	case err := <-errChannel:
+		return nil, err
+	case res := <-resChannel:
+		return res, nil
+	}
+}
+
+// ReadDir is similar to ioutil.ReadDir, but returns an error after a timeout.
+func ReadDir(path string) ([]os.FileInfo, error) {
+	return internalReadDirTimeout(path, ioTimeout)
+}
+
+func internalReadDirTimeout(path string, timeout time.Duration) ([]os.FileInfo, error) {
+	timer := time.NewTimer(timeout)
+	resChannel := make(chan []os.FileInfo)
+	errChannel := make(chan error)
+
+	defer timer.Stop()
+
+	go func() {
+		entries, err := ioutil.ReadDir(path)
+		if err != nil {
+			errChannel <- err
+		} else {
+			resChannel <- entries
+		}
+	}()
+
+	select {
+	case <-timer.C:
+		return nil, fmt.Errorf("ReadDir on %v timed out after %v", path, timeout)
+	case err := <-errChannel:
+		return nil, err
+	case res := <-resChannel:
+		return res, nil
+	}
+}
```

Comment:
internalReadDirTimeout() and internalLstatTimeout() almost identical except the actual operation. I am wondering whether we could have a function wrapper which can provide support for any operations that need timeout.